modules/local/locidex_select.nf

```diff
@@ -0,0 +1,194 @@
+
+
+Locidex Manifest is setup as:
+    {
+        db_name: [
+            {
+                "path": path/to/db/relative/to/manifest
+                # DB Config data, the newest db data will be selected as versions are not standardized
+                "config": {
+                    "db_name": "Locidex Database 1",
+                    "db_version": "1.0.0",
+                    "db_date": "yyyy-MM-dd",
+                    "db_author": "test1",
+                    "db_desc": "test1",
+                    "db_num_seqs": 53,
+                    "is_nucl": true,
+                    "is_prot": true,
+                    "nucleotide_db_name": "nucleotide",
+                    "protein_db_name": "protein"
+                }
+            }
+        ]
+    }
+*/
+
+import groovy.json.JsonSlurper
+import groovy.json.JsonBuilder
+import java.text.SimpleDateFormat
+
+
+process LOCIDEX_SELECT {
+    tag "$meta.id"
+    label "process_single"
+
+    input:
+    tuple val(meta), val(top_hit), val(contigs)
+    // TODO check that manifest is copied propely in azure, path identifier does not stage file
+    val manifest // This is a json file to be parsed
+
+    output:
+    tuple val(meta), val(contigs), val(scheme), val(paired_p), emit: db_data
+    tuple val(meta), path(output_config), emit: config_data
+
+    exec:
+    if(params.allele_scheme == null && params.locidex.allele_database == null){
+        exit 1, "Allele calling is enabled but there is no allele scheme or locidex allele database location present."
+    }
+
+    // Tokenize the "top_hit" or species value to identify all relevant match parts of the string
+    def species_data = top_hit.split('_|\s')
+    species_data = species_data*.toLowerCase()
+
+    // De-serialize the manifest file from the database location
+    def jsonSlurper = new JsonSlurper()
+    String json_data = manifest.text
+    def allele_db_data = jsonSlurper.parseText(json_data)
+    def allele_db_keys = allele_db_data.keySet() as String[]
+
+    // Tokenize all database keys for lookup of species top hit in the database names
+    def databases = []
+    def shortest_entry = Integer.MAX_VALUE
+    def idx = 0
+    for(i in allele_db_keys){
+        def db_tokens = i.split('_|\s')
+        for(g in db_tokens){
+            def tok_size = g.size()
+            if(tok_size < shortest_entry){
+                shortest_entry = tok_size
+            }
+        }
+        databases.add(new Tuple(db_tokens*.toLowerCase(), i, idx))
+        idx += 1
+    }
+    def db_tokes_pos = 0
+    def db_key = 1
+    def db_idx_pos = 2
+
+    // Remove spurious characters from tokenized string
+    species_data = species_data.findAll { it.size() >= shortest_entry }
+
+    // A default locidex database is set to null as there should be no option set
+    // a default database can be set, but this process will then be skipped
+    def db_opt = null
+
+
+    paired_p = false // Sets predicate for db identification as false by default
```

Comment:
I just noticed the unresolved TODO in this comment. I think I need to verify that the optimal match is found through a test first then I can make this change.

I was declaring db_found locally in scope so that if I continue to iterate through the options for a "best match" that no state affects finding the correct database.

I think I need to add a test case here to verify the correct behaviour of the database finding, then I can will make this suggested change. as you are right it is not doing anything currently

Reply:
Another thought - could you use the `find` function instead - it will have downstream implications though....

Something like this....
`def matchedDb = databases.find { db ->
    def match_size = db[DB_TOKES_POS].size()
    def tokens = tokenize_values(species_data, match_size)
    compare_lists(db[DB_TOKES_POS], tokens)
}`

`if (matchedDb) {
    def selected_db = select_locidex_db_path(allele_db_data[matchedDb[DB_KEY]], matchedDb[DB_KEY])
    write_config_data(selected_db, output_config)
    scheme = join_database_paths(selected_db)
} else {
    write_config_data(["No database selected."], output_config)
}`

Reply:
I made some changes to how things are found. and I performed something similar to what you suggested! I used a sort instead of find so that I could perform additional checks later on. I might be able to use a `findAll` but I think I would still need to sort anyways.

Also this comment did inspire the changes in the code however!

Reply:
fixed in: https://github.com/phac-nml/mikrokondo/pull/96/commits/2fb4ac947e3888de3cfcc6f27b83903719999797